### PR TITLE
Feat: Issue #744: Curl_trait created, connector curl separated and fl…

### LINKED ISSUE #744: Create a curl trait, separate from the connector curl to be able to create a curl_flow
Screenshot from 2023-05-11 16-46-56

image

### PR DESCRIPTION
**Pull request by marcghaly**
Resolves #744

Hi @keevan,

This PR is about splitting the curl connector to have a trait reusable from a flow curl step,

Regards,

Marc

### CODE REVIEW

**Review by keevan — COMMENTED** [review]
Looks mostly fine. I've added a comment. Also, could we also add a version bump please.

Thanks Marc

File: classes/local/step/curl_trait.php [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+        // TODO: It would be good to define and list any fixed but exposed
+        // fields which the user can use and map to on the edit page.
+        $variables->set('response', (object) [
+            'result' => $result,
+            'info' => (object) $info,
+            'destination' => $destination,
+        ]);
+
+        return true;
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply classes/local/step/curl_trait.php]: Thanks for catching that !

**Comment by marcghaly** [comment]
Hi @keevan,

Fixed for the returned value in the trait is fixed and version bump has been added.

Thank you !

Regards,

Marc

**Comment by marcghaly** [comment]
Also, I think it would be should to be able to reuse keys/values from previous inputs by evaluating them, what do you think ?

**Review by keevan — APPROVED** [review]
Looks good to me

**Comment by keevan** [comment]
> Also, I think it would be should to be able to reuse keys/values from previous inputs by evaluating them, what do you think ?

I'm not sure what you mean. Each flow step uses the previous step's input (e.g. record). 

If that's not what you mean, can you clarify with an example please?

**Comment by marcghaly** [comment]
Hi @keevan,

I have added a commit over maybe this will be more explanatory ? Maybe I need to drop it, I am not sure if I can use the following syntax `record.Somekey` ?

Regards,

Marc

**Comment by marcghaly** [comment]
Double checked and can use the `record` syntax, Thank you Kevin

File: classes/local/step/curl_trait.php [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+            $dbgcommand .= ' --output ' . helper::bash_escape($config->destination);
+        }
+
+        // Log the raw curl command.
+        $this->enginestep->log($dbgcommand);
+        $variables->set('dbgcommand', $dbgcommand);
+
+        // We do not need to go any further if curl is not going to be called.
+        if ($this->enginestep->engine->isdryrun && $this->has_side_effect()) {
+            return true;
```
Comment (keevan) [inline classes/local/step/curl_trait.php:281]: Missing return `$input` here?

**Pull request merged**